ERROR: Ran out of tokens

On any example HolyC program I've tried, when I go to compile it I will get an: 

`ERROR: line 7: Ran out of tokens`

on an obscure line number. 

The program:

```
U0 Function(U8 *name="Bob")
{
  "%s\n",name;
}

Function; 
Function("Terry");   
```

Any idea why this happens? Am I doing something wrong? How can I fix it?

Another issue worth noting:

When running the unit tests:

It initially didn't work and I had to install sqlite-devel, then the initial tests started to pass but it gets hung at 'Test - Simple Threading'

```
[parker@jarvis holyc-lang]$ make unit-test
make -C ./build unit-test
make[1]: Entering directory '/home/parker/holyc-lang/build'
[ 94%] Built target hcc
[100%] Building and running unit tests
Running unit tests...
Testing if else
PASSED: 4/4
=====
Test Assigning pointer to a pointer from a class, with brackets: PASSED: 1/1
====
Test Assigning pointer to a pointer from a class, no brackets: PASSED: 1/1
====
Test - Juggling an array MAlloc(I64 * 10): PASSED: 4/4
====
Test - Pointer to a pointer from a class: PASSED: 1/1
=====
Test - Taking the address of a local: PASSED: 1/1
Test - class definition: PASSED: 6/6
====
Test - Linked list recusive definition class: PASSED: 5/5
====
Test - 'break' and 'continue' inside while and for loops: PASSED: 4/4
====
Test - Default function parameters: PASSED: 1/1
====
Test - qsort, recursion and default function params: PASSED: 1/1
====
Assignment tests:
= test: PASSED: 2/2
+= test: PASSED: 2/2
-= test: PASSED: 2/2
*= test: PASSED: 2/2
/= test: PASSED: 2/2
%= test: PASSED: 2/2
&= test: PASSED: 2/2
|= test: PASSED: 2/2
^= test: PASSED: 2/2
>>= test: PASSED: 2/2
<<= test: PASSED: 2/2
=====
Test - Function pointers: PASSED: 5/5
====
Test - Simple Threading: 

```

After hittitng CTRL-C it skips the test and the rest seem to pass:

But it would hang there forever if I didn't. I wonder if this is a related to the token error problem. 🤔

```
^CFAILED to run: 13_threads.HC
Test Mathematics:
Subtraction - : PASSED: 7/7
Multiplication * : PASSED: 5/5
Modulo % : PASSED: 5/5
Addition + : PASSED: 5/5
Division / : PASSED: 3/3
MATHS_SHL << : PASSED: 3/3
MATHS_SHR >> : PASSED: 3/3
Xor ^: PASSED: 3/3
Or |: PASSED: 3/3
And &: PASSED: 3/3
Not ~: PASSED: 3/3
=====
Test - MemCpy, MemSet: PASSED: 4/4
Test - comma separated declarators & types: PASSED: 3/3
====
Test - Variable arguments: PASSED: 15/15
=====
Test - Type casting:
PASSED: 9/9
=====
Test - Switch statement: PASSED: 7/7
Loop test suite
Testing For Loop: PASSED: 2/2
Testing while loop: PASSED: 1/1
Testing do-while loop: PASSED: 1/1
=====
Testing GOTO: PASSED: 2/2
=====
Addressing: PASSED: 2/2
====
Malloced Arrays:
PASSED: 10/10
PASSED: 2/2
=====
Test: Union and class inheritance: PASSED: 4/4
====
Test Scripting: Test Scripting: PASSED: 4/4
===
Test - Floating point arithmetic:
PASSED: 5/5
====
Test - auto keyword: PASSED: 2/2
====
Test - Arrays & multi dimensional: PASSED: 4/4
====
Test - Preprocessor: PASSED: 6/6
====
Test - FzF fuzzy finding: PASSED: 7/7
====
Test - JSON parsing: {"hello": 10,"key": "some string","id": null,"bool2": false,"bool": true,"array": [1,30030303,430,-14]}
PASSED: 10/10
====
Test - sqlite bindings & abstraction: 
PASSED: 4/4
====
Test - local static variables and global variables: 
PASSED: 3/3
====
Test - Global and static array
PASSED: 2/2
====
Test - basic io: PASSED: 4/4
====
make[3]: *** [CMakeFiles/Makefile2:110: CMakeFiles/unit-test.dir/all] Interrupt
make[2]: *** [CMakeFiles/Makefile2:117: CMakeFiles/unit-test.dir/rule] Interrupt
make[1]: *** [Makefile:182: unit-test] Interrupt
make: *** [Makefile:18: unit-test] Interrupt
```

--- a/Makefile
+++ b/Makefile
@@ -5,6 +5,16 @@ default: all
 
 .PHONY: all
 
+# To add sqlite3 support add -DHCC_LINK_SQLITE3=1 to the below like so:
+#```
+#all:
+#	cmake -S ./src -B ./build -G 'Unix Makefiles' \
+#		-DCMAKE_C_COMPILER=$(_C_COMPILER) \
+#		-DCMAKE_BUILD_TYPE=$(_BUILD_TYPE) \
+#		-DHCC_LINK_SQLITE3=1 \
+#		&& $(MAKE) -C ./build -j2
+#```
+
 all:
 	cmake -S ./src -B ./build -G 'Unix Makefiles' \
 		-DCMAKE_C_COMPILER=$(_C_COMPILER) \

--- a/README.md
+++ b/README.md
@@ -92,6 +92,9 @@ make -C ./build install
 ```
 This will install the compiler and holyc libraries for strings, hashtables, 
 I/O, maths, networking, JSON parsing etc... see ./src/holyc-lib/.
+
+If you would like to include `sqlite3` then please add `-DHCC_LINK_SQLITE3=1`
+to either the Makefile or when configuring cmake.
  
 ## Using the compiler
 Once the compiler has been compiled the following options are available, they 

--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -20,6 +20,10 @@ if(CMAKE_BUILD_TYPE STREQUAL "Debug")
     set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -fsanitize=address")
 endif()
 
+if (HCC_LINK_SQLITE3)
+    add_compile_definitions(HCC_LINK_SQLITE3)
+endif()
+
 # Optionally, add more flags or options depending on the build type
 # For example, optimization flags for Release builds:
 set(CMAKE_C_FLAGS_RELEASE "${CMAKE_C_FLAGS_RELEASE} -O2")
@@ -86,4 +90,8 @@ install(TARGETS hcc DESTINATION ${CMAKE_INSTALL_PREFIX}/bin)
 install(FILES ${CMAKE_SOURCE_DIR}/holyc-lib/tos.HH DESTINATION ${CMAKE_INSTALL_PREFIX}/include)
 
 # Custom installation step to run `hcc` during install
+#if (DEFINED HCC_LINK_SQLITE3)
+#    install(CODE "execute_process(COMMAND hcc -DHCC_LINK_SQLITE3 -lib tos ${CMAKE_SOURCE_DIR}/holyc-lib/all.HC WORKING_DIRECTORY ${CMAKE_SOURCE_DIR}/holyc-lib)")
+#else()
 install(CODE "execute_process(COMMAND hcc -lib tos ${CMAKE_SOURCE_DIR}/holyc-lib/all.HC WORKING_DIRECTORY ${CMAKE_SOURCE_DIR}/holyc-lib)")
+    #endif()

--- a/src/cctrl.c
+++ b/src/cctrl.c
@@ -117,6 +117,11 @@ static void cctrlAddBuiltinMacros(Cctrl *cc) {
     le->tk_type = TK_STR;
     dictSet(cc->macro_defs,"__TIMESTAMP__",le);
 
+#ifdef HCC_LINK_SQLITE3
+    le = lexemeSentinal();
+    dictSet(cc->macro_defs,"__HCC_LINK_SQLITE3__",le);
+#endif
+
     len = snprintf(version,bufsize,"v0.0.2-alpha");
     version[len] = '\0';
     le = lexemeNew(version,len);

--- a/src/holyc-lib/all.HC
+++ b/src/holyc-lib/all.HC
@@ -14,4 +14,6 @@
 #include "./fzf.HC"
 #include "./json.HC"
 #include "./net.HC"
-#include "./sqllite.HC"
+#ifdef __HCC_LINK_SQLITE3__ 
+  #include "./sqllite.HC"
+#endif

--- a/src/main.c
+++ b/src/main.c
@@ -21,6 +21,14 @@
 #define LIB_PATH "/usr/local/lib"
 #define LIB_BUFSIZ 256
 
+#define CLIBS_BASE "-lpthread -ltos -lc -lm"
+
+#ifdef HCC_LINK_SQLITE3
+#define CLIBS CLIBS_BASE" -lsqlite3"
+#else
+#define CLIBS CLIBS_BASE
+#endif
+
 typedef struct hccOpts {
     int print_ast;
     int print_tokens;
@@ -58,7 +66,7 @@ typedef struct hccLib {
 } hccLib;
 
 char *getVersion(void) {
-    return "0.0.1";
+    return "0.0.2";
 }
 
 int hccLibInit(hccLib *lib, hccOpts *opts, char *name) { 
@@ -71,7 +79,7 @@ int hccLibInit(hccLib *lib, hccOpts *opts, char *name) {
     snprintf(lib->dylib_name,LIB_BUFSIZ,"%s.dylib",name);
     snprintf(lib->dylib_version_name,LIB_BUFSIZ,"%s.0.0.1.dylib",name);
     aoStrCatPrintf(
-            dylibcmd, "cc -dynamiclib -Wl,-install_name,%s/%s -o %s -lpthread -lsqlite3 -lc -lm",
+            dylibcmd, "cc -dynamiclib -Wl,-install_name,%s/%s -o %s "CLIBS,
             LIB_PATH, name, lib->dylib_version_name,lib->dylib_name);
     aoStrCatPrintf(installcmd,
             "cp -pPR ./%s /usr/local/lib/lib%s && "
@@ -85,7 +93,7 @@ int hccLibInit(hccLib *lib, hccOpts *opts, char *name) {
     snprintf(lib->dylib_name,LIB_BUFSIZ,"%s.so",name);
     snprintf(lib->dylib_version_name,LIB_BUFSIZ,"%s.so.0.0.1",name);
     aoStrCatPrintf(
-            dylibcmd, "gcc -fPIC -shared -Wl,-soname,%s/%s -o %s -lpthread -lsqlite3 -lc -lm",
+            dylibcmd, "gcc -fPIC -shared -Wl,-soname,%s/%s -o %s "CLIBS,
             LIB_PATH,name,lib->dylib_name,lib->dylib_name);
     aoStrCatPrintf(installcmd,
             "cp -pPR ./%s /usr/local/lib/lib%s",
@@ -200,7 +208,7 @@ void emitFile(aoStr *asmbuf, hccOpts *opts) {
     hccLib lib;
     if (opts->emit_object) {
         writeAsmToTmp(asmbuf);
-        aoStrCatPrintf(cmd, "gcc -c %s -lpthread -ltos -lsqlite3 -lc -lm %s -o ./%s",
+        aoStrCatPrintf(cmd, "gcc -c %s "CLIBS" %s -o ./%s",
                 ASM_TMP_FILE,opts->clibs,opts->obj_outfile);
         system(cmd->data);
     } else if (opts->asm_outfile && opts->assemble_only) {
@@ -227,10 +235,10 @@ void emitFile(aoStr *asmbuf, hccOpts *opts) {
     } else {
         writeAsmToTmp(asmbuf);
         if (opts->clibs) {
-            aoStrCatPrintf(cmd, "gcc -L/usr/local/lib %s -lpthread -ltos -lsqlite3 %s -lc -lm -o %s", 
+            aoStrCatPrintf(cmd, "gcc -L/usr/local/lib %s %s "CLIBS" -o %s", 
                     ASM_TMP_FILE,opts->clibs,opts->output_filename);
         } else {
-            aoStrCatPrintf(cmd, "gcc -L/usr/local/lib %s -lpthread -ltos -lsqlite3 -lc -lm -o %s", 
+            aoStrCatPrintf(cmd, "gcc -L/usr/local/lib %s "CLIBS" -o %s", 
                     ASM_TMP_FILE, opts->output_filename);
         }
         system(cmd->data);
@@ -251,7 +259,7 @@ void emitFile(aoStr *asmbuf, hccOpts *opts) {
 
 void usage(void) {
     fprintf(stderr,
-            "HolyC Compiler 2024. UNSTABLE\n"
+            "HolyC Compiler 2024. %s UNSTABLE\n"
             "hcc [..OPTIONS] <..file>\n\n"
             "OPTIONS:\n"
             "  -ast     Print the ast and exit\n"
@@ -267,7 +275,8 @@ void usage(void) {
             "  -run     Immediately run the file (not JIT)\n"
             "  -g       Not implemented\n"
             "  -D<var>  Set a compiler #define (does not accept a value)\n"
-            "  --help   Print this message\n");
+            "  --help   Print this message\n",
+            getVersion());
     exit(1);
 }
 

--- a/src/tests/run.HC
+++ b/src/tests/run.HC
@@ -81,7 +81,15 @@ U0 QSortStrings(U8 **arr, I64 high, I64 low=0)
 
 Bool ShouldExclude(U8 *filename, I64 namelen)
 {
-  U8 *exclude[] = {"run.HC", "testhelper.HC", "tests.sh", "Makefile", "a.out",
+  U8 *exclude[] = {
+    "run.HC",
+    "testhelper.HC",
+    "tests.sh",
+    "Makefile",
+    "a.out",
+#ifndef __HCC_LINK_SQLITE3__
+    "32_sql.HC",
+#endif
     "test-runner"};
   auto len = sizeof(exclude)/sizeof(exclude[0]);
   for (auto i = 0; i < len; ++i) {

